test/distributed/pipelining/test_schedule.py

```diff
@@ -344,7 +362,190 @@ def test_grad_with_manual_interleaved(self, ScheduleClass):
 
 instantiate_parametrized_tests(ScheduleTest)
 
+
+def format_pipeline_order(pipeline_order: Dict[int, List[Optional[_Action]]]):
+    import itertools
+
+    column_labels = ["Step " + str(i) for i in range(len(pipeline_order[0]))]
+    # Sorting the dictionary by keys and retrieving values in that order
+    rank_actions = [pipeline_order[key] for key in sorted(pipeline_order)]
+    pipeline_order_with_labels = [column_labels] + rank_actions
+    # Calculate the maximum length of each column
+    max_lengths = [
+        max(len(str(item)) if item is not None else 0 for item in col)
+        for col in itertools.zip_longest(*pipeline_order_with_labels, fillvalue="")
+    ]
+    # Format each row
+    formatted_rows = [
+        " ".join(f"{str(item):<{max_lengths[i]}}" for i, item in enumerate(row))
+        for row in pipeline_order_with_labels
+    ]
+    # Join the rows into a single string
+    formatted_table = (
+        "=========== ALL_RANK_ACTIONS ===========\n" + "\n".join(formatted_rows) + "\n"
+    )
+    return formatted_table
+
+
+class TestSchedulePlan(unittest.TestCase):
+    def _validate_pipeline_order(
+        self,
+        pipeline_order: Dict[int, List[Optional[_Action]]],
+        num_microbatches: int,
+        num_stages: int,
+    ):
+        """
+        pipeline_order[rank] = [(computation_type, microbatch_index, stage_index), ...]
+
+        Validating that the pipeline order follows the rules:
```

Comment:
this seems like a good start. I think it's a good idea to have some 'logic verification' tests for our schedules.

However, it would be more interesting if it could check the send/recv ops including batching.  Somehow we have to figure out if we are doing the right comms or not, without actually running on gpus.

Reply:
I think that when the schedule logic is determined at runtime and you have to figure out what when to call `get_fwd_recv_ops`, `get_fwd_send_ops`, etc. then checking batching is a lot more helpful.

For these new schedules I found that there wasn't really an issue with figuring out how operations were batched since the logic based on each step is straightforward. At every time step we were guaranteed to have each batched op across ranks match with each other and if they didn't it was pretty easy to tell. Usually the issue was because a stage was too early or too late in the plan, which this test did help a lot to verify as I was developing.

But yeah it is fair that the schedules are not "actually being run" with this test so we can't say for absolute certainty that it works. If we were able to check send/recv ops that might help increase the confidence, but I am pretty confident that the current `step_microbatches` implementation is good and our current tests can catch issues with it, which in turns means the plans should "just work"

Reply:
mock the forward / backward.

Check the results of get_recv ops and get_send_ops. Validate these

Reply:
https://github.com/pytorch/pytorch/issues/127965